Smart pointers in GPUProcess/graphics

squelart opened the pull request:
#### 2f5f9c0d21b50540d40af7e2bccf23ff2fcfa4ed
<pre>
Smart pointers in GPUProcess/graphics
<a href="https://bugs.webkit.org/show_bug.cgi?id=262465">https://bugs.webkit.org/show_bug.cgi?id=262465</a>
rdar://problem/116322521

Reviewed by Chris Dumez.

Adding smart pointers as expected by clang static analyzer.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBuffer):

Canonical link: <a href="https://commits.webkit.org/268807@main">https://commits.webkit.org/268807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12964ab7c124467ada1ca05539bd4c2f46b5560d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20591 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21009 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21661 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22481 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19217 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24263 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21187 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20576 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20812 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20650 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17901 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23338 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17816 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24979 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18895 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18885 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22919 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19466 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16544 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18686 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4973 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23022 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19285 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
```diff
@@ -280,9 +280,9 @@ NO_RETURN_DUE_TO_ASSERT void DrawFilteredImageBuffer::apply(GraphicsContext&) co
 }
 
-void DrawFilteredImageBuffer::apply(GraphicsContext& context, ImageBuffer* sourceImage, FilterResults& results)
+void DrawFilteredImageBuffer::apply(GraphicsContext& context, RefPtr<ImageBuffer> sourceImage, FilterResults& results)
```
This is a weird change. As far as I know, we're not supposed to change parameter types during this smart pointer effort, unless there is a good reason to. Here we're just calling `.get()` on the smart pointer so no good reason I can see.

squelart replied:
Ah yes, I was told about that, but that patch was already in-flight so I forgot to revert this, will do...

cdumez commented on Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
```diff
@@ -286,7 +286,7 @@ void RemoteDisplayListRecorder::drawFilteredImageBuffer(std::optional<RenderingR
 
     RefPtr cachedFilter = resourceCache().cachedFilter(filter->renderingResourceIdentifier());
-    auto* cachedSVGFilter = dynamicDowncast<SVGFilter>(cachedFilter.get());
+    RefPtr cachedSVGFilter = dynamicDowncast<SVGFilter>(cachedFilter);
```
`WTFMove(cachedFilter)` could be more efficient here.

cdumez commented on Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
```diff
@@ -339,7 +339,7 @@ void RemoteDisplayListRecorder::drawImageBuffer(RenderingResourceIdentifier imag
     }
 
-    handleItem(DisplayList::DrawImageBuffer(imageBufferIdentifier, destinationRect, srcRect, options), *sourceImage);
+    handleItem(DisplayList::DrawImageBuffer(imageBufferIdentifier, destinationRect, srcRect, options), sourceImage.releaseNonNull());
```
I am assuming all these are because of the change to apply() however, they don't seem like an improvement to me. It doesn't seem to avoid ref-counting churn so it is not clear why we're doing this.

cdumez reviewed (approved):
LGTM

squelart committed:
Smart pointers in GPUProcess/graphics
https://bugs.webkit.org/show_bug.cgi?id=262465
rdar://problem/116322521

Reviewed by Chris Dumez.

Adding smart pointers as expected by clang static analyzer.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::drawFilteredImageBuffer):

Canonical link: https://commits.webkit.org/268807@main
